### DISCUSSION TITLE
Upgrade - axis-cgi/firmwaremanagement.cgi

### DISCUSSION THREAD

**Discussion by dmannion1972-zz** [AxisCommunications/discussions#50] [VAPIX] [answered]
I am using Python, to perform firmware upgrades but the following code results in a 200 status code and an error of "message": "JSON request is too large."  

import requests
import json
from requests.auth import HTTPDigestAuth
cam_user = 'root'
cam_pass = 'pass'
auth=HTTPDigestAuth(cam_user, cam_pass)

url = "http://192.168.30.14/axis-cgi/firmwaremanagement.cgi?"

files = {'file': ('P1468.bin',open('P1468.bin', 'rb'),'application/octet-stream')}
payload = {
  "apiVersion": "1.3",
  "context": "abc",
  "method": "upgrade"
}
headers = {
  'Content-Type': 'application/json'
}

response = requests.post(url, auth=auth,headers=headers,data=payload,files=files)
print(response.status_code)
print(response.text)

Using the firmware status CGI I have no issues but upgrade CGI is not working so well. Thoughts?

**Comment by dmannion1972-zz** [comment, marked as answer]
Figured it out.  Here is the python code that worked.  

import requests
import json
from requests.auth import HTTPDigestAuth
cam_user = 'root'
cam_pass = 'pass'
auth=HTTPDigestAuth(cam_user, cam_pass)

url = "http://192.168.30.14/axis-cgi/firmwaremanagement.cgi?"

payload = {
  "apiVersion": "1.3",
  "context": "abc",
  "method": "upgrade"
}

files = {
  'file': [('file.bin',open('P1468.bin', 'rb'),'application/octet-stream')],
  'payload': json.dumps(payload)
  }

fields={'payload': json.dumps(payload)}

response = requests.post(url, auth=auth,data=fields,files =[('file',open('P1468.bin', 'rb'))])

print(response.status_code)
print(response.text)

**Reply by wiener-dog** [reply]
I have 100's of cameras to setup/update.  I think I can use this to update, put it in a wrapper and feed it ip addresses.  Now.. if I could just get the pwdgrp.cgi interface to work, to setup the intial root account.  That's black magic...

**Reply by gdmannion** [reply]
If it helps, here is an example code using Python.  It either creates a new user for a brand new device/factory defaulted device or creates a user that already has an admin account created.  

import requests
from requests.auth import HTTPDigestAuth
from getpass import getpass

def select_operation():
    print("Welcome to the user creation python script")
    print("1 - This script is intended to help create an admin user for a new product or factory default device")
    print("2 - You can also use this script to add a user to an already configured Axis device")
    selec_op = input("Please select option 1 or option 2: ")
    return selec_op

# Function to create a new user
def create_new_user(ip_address):
    # Prompt the user for the admin username and password for authentication.
    admin_username = input("Enter admin username: ")
    admin_password = getpass("Enter admin password: ")

    # Prompt the user for the new user's username and password.
    user_add = input("Enter username you would like to create: ")
    password_add = getpass("Enter password for the new user: ")

    # Define user groups
    adminuser = "grp=users&sgrp=admin%3Aoperator%3Aviewer%3Aptz"

    # Set up HTTP Digest Authentication using the admin credentials
    auth = HTTPDigestAuth(admin_username, admin_password)

    # Construct the URL for creating the new user
    user_url = f"http://{ip_address}/axis-cgi/pwdgrp.cgi?action=add&user={user_add}&pwd={password_add}&grp=users&sgrp={adminuser}"

    # Send the request to create the user
    user_response = requests.get(user_url, auth=auth)

    # Check if the user was created successfully
    if user_response.status_code == 200:
        print("User created")
        url = f"http://{ip_address}/axis-cgi/pwdgrp.cgi?action=get"
        response = requests.get(url, auth=auth)
        print("User Information:")
        print(response.text)
    else:
        print("Failed to create user")
        print("Response Text:", user_response.text)
        print("Status Code:", user_response.status_code)

# Function to add a user to an existing device
def add_user(ip_address):
    # Prompt the user for the admin username and password for authentication.
    admin_username = input("Enter admin username: ")
    admin_password = getpass("Enter admin password: ")

    # Prompt the user for the new user's username and password.
    user_add = input("Enter username you would like to create: ")
    password_add = getpass("Enter password for the new user: ")

    # Define user groups
    adminuser = "grp=users&sgrp=admin%3Aoperator%3Aviewer%3Aptz"
    operatoruser = "grp=viewer%3Aoperator"
    viewuser = "grp=viewer%3Aoperator%3Aadmin%3Aptz"

    # Select user permissions
    selectuser = input("Select User Permissions: 1 for admin, 2 for operator, 3 for viewer: ")

    # Set up HTTP Digest Authentication using the admin credentials
    auth = HTTPDigestAuth(admin_username, admin_password)

    # Determine the user group based on the selected permissions
    if selectuser == "1":
        user_url = adminuser
    elif selectuser == "2":
        user_url = operatoruser
    elif selectuser == "3":
        user_url = viewuser
    else:
        print("Invalid selection. Please select 1, 2, or 3.")
        return

    # Construct the URL for adding the user
    user_url = f"http://{ip_address}/axis-cgi/pwdgrp.cgi?action=add&user={user_add}&pwd={password_add}&grp=users&sgrp={user_url}"

    # Send the request to add the user
    user_response = requests.get(user_url, auth=auth)

    # Check if the user was added successfully
    if user_response.status_code == 200:
        print("User created")
        url = f"http://{ip_address}/axis-cgi/pwdgrp.cgi?action=get"
        response = requests.get(url, auth=auth)
        print("User Information:")
        print(response.text)
    else:
        print("Failed to create user")
        print("Response Text:", user_response.text)
        print("Status Code:", user_response.status_code)

# Prompt the user for the camera's IP address.
ip_address = input("Enter the IP address of the camera: ")

# Main loop to create or add users
create_another = "yes"
while create_another.lower() == "yes":
    option = select_operation()
    if option == "1":
        create_new_user(ip_address)
    elif option == "2":
        add_user(ip_address)
    else:
        print("Invalid option selected.")
    create_another = input("Do you want to create or add another user? (yes/no): ")

print("Exiting...")

**Comment by osbuassi** [comment]
Hello,
I tried this code (python 3.9.5 64-bit running over Visual Studio Code ), but I have some trouble with it. 

**A) files Argument data type:**
_error:
Argument of type "dict[str, list[tuple[Unknown, BufferedReader, Literal['application/octet-stream']]]]" cannot be assigned to parameter "files" of type "_Files | None" in function "post"
  Type "dict[str, list[tuple[Unknown, BufferedReader, Literal['application/octet-stream']]]]" cannot be assigned to type "_Files | None"
    "dict[str, list[tuple[Unknown, BufferedReader, Literal['application/octet-stream']]]]" is incompatible with "Mapping[str, _FileSpec]"
      Type parameter "_VT_co@Mapping" is covariant, but "list[tuple[Unknown, BufferedReader, Literal['application/octet-stream']]]" is not a subtype of "_FileSpec"
        Type "list[tuple[Unknown, BufferedReader, Literal['application/octet-stream']]]" cannot be assigned to type "_FileSpec"
          "list[tuple[Unknown, BufferedReader, Literal['application/octet-stream']]]" is incompatible with protocol "SupportsRead[str | bytes]"
          "list[tuple[Unknown, BufferedReader, Literal['application/octet-stream']]]" is incompatible with "str"
          "list[tuple[Unknown, BufferedReader, Literal['application/octet-stream']]]" is incompatible with "bytes"
          "list[tuple[Unknown, BufferedReader, Literal['application/octet-stream']]]" is incompatible with "_FileSpecTuple2"_

**B) I changed my request using "files =[('file',open('P1468.bin', 'rb'))]" and I get this error:**
_Traceback (most recent call last):
  File "C:\Users\Documents\Help-IT DATA\BMW\Scripts macros y programas\VAPIX get and post\fimrware_version.py", line 90, in <module>
    firmware_act (usr, pswd, cam_ip, filepath)
  File "C:\Users\Documents\Help-IT DATA\BMW\Scripts macros y programas\VAPIX get and post\fimrware_version.py", line 43, in firmware_act
    req_FW = requests.post(url, auth=HTTPDigestAuth(usr, pswd),files=file)
  File "C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\requests\api.py", line 115, in post
    return request("post", url, data=data, json=json, **kwargs)
  File "C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\requests\api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
  File "C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\requests\sessions.py", line 575, in request
    prep = self.prepare_request(req)
  File "C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\requests\sessions.py", line 486, in prepare_request
    p.prepare(
  File "C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\requests\models.py", line 371, in prepare
    self.prepare_body(data, files, json)
  File "C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\requests\models.py", line 556, in prepare_body
    (body, content_type) = self._encode_files(files, data)
  File "C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\requests\models.py", line 183, in _encode_files
    fn, fp, ft, fh = v
ValueError: not enough values to unpack (expected 4, got 1)_

**C) files varible is not used in request** 

I can already get Firmware version using 'method':'status'

thanks for reading

**Reply by ghost** [reply]
Do you have the bin file for the Axis product? In my example, I was doing this using p1468 (P1468.bin), but you need to download the correct firmware file for your product, and the file must be in the same folder as your script.

**Reply by ghost** [reply]
Here is a demo video showing the upgrade https://youtu.be/sY59GjZKuDo and examples of the script https://github.com/dmannion1972/firmwaremanagement_upi_upgrade

**Reply by osbuassi** [reply]
Thank you for help me with this. The second script you posted in GitHub works. 
[fwupgrade_inputnamefile,py](https://github.com/dmannion1972/firmwaremanagement_upi_upgrade/blob/main/fwupgrade_inputnamefile%2Cpy)
